scripts/Muon/GUI/MuonAnalysis/loadfile/load_file_view.py

```diff
@@ -0,0 +1,129 @@
+
+    def __init__(self, parent=None):
+        super(BrowseFileWidgetView, self).__init__(parent)
+
+        self.horizontal_layout = None
+        self.browse_button = None
+        self.file_path_edit = None
+
+        self.setup_interface_layout()
+
+        self._store_edit_text = False
+        self._stored_edit_text = ""
+        self._cached_text = ""
+
+    def setup_interface_layout(self):
+        self.setObjectName("BrowseFileWidget")
+        self.resize(500, 100)
+
+        self.browse_button = QtGui.QPushButton(self)
+
+        size_policy = QtGui.QSizePolicy(QtGui.QSizePolicy.Fixed, QtGui.QSizePolicy.Fixed)
+        size_policy.setHorizontalStretch(0)
+        size_policy.setVerticalStretch(0)
+        size_policy.setHeightForWidth(self.browse_button.sizePolicy().hasHeightForWidth())
+
+        self.browse_button.setSizePolicy(size_policy)
+        self.browse_button.setMinimumSize(QtCore.QSize(100, 50))
+        self.browse_button.setObjectName("browseButton")
+        self.browse_button.setText("Browse")
+
+        self.file_path_edit = QtGui.QLineEdit(self)
+
+        size_policy = QtGui.QSizePolicy(QtGui.QSizePolicy.Minimum, QtGui.QSizePolicy.Fixed)
+        size_policy.setHorizontalStretch(0)
+        size_policy.setVerticalStretch(0)
+        size_policy.setHeightForWidth(self.file_path_edit.sizePolicy().hasHeightForWidth())
+
+        self.file_path_edit.setSizePolicy(size_policy)
+        self.file_path_edit.setMinimumSize(QtCore.QSize(200, 50))
+        self.file_path_edit.setToolTip("")
+        self.file_path_edit.setObjectName("filePathEdit")
+
+        self.horizontal_layout = QtGui.QHBoxLayout(self)
+        self.horizontal_layout.setObjectName("horizontalLayout")
+        self.horizontal_layout.addWidget(self.file_path_edit)
+        self.horizontal_layout.addWidget(self.browse_button)
+
+        self.setLayout(self.horizontal_layout)
+
+    def on_browse_clicked(self, slot):
+        self.browse_button.clicked.connect(slot)
+
+    def on_file_edit_changed(self, slot):
+        self.file_path_edit.returnPressed.connect(slot)
+
+    def show_file_browser_and_return_selection(self, file_filter, search_directories, multiple_files=False):
+        default_directory = search_directories[0]
+        if multiple_files:
+            chosen_files = QtGui.QFileDialog.getOpenFileNames(self, "Select files", default_directory,
+                                                              file_filter)
+            return [str(chosen_file) for chosen_file in chosen_files]
+        else:
+            chosen_file = QtGui.QFileDialog.getOpenFileName(self, "Select file", default_directory,
+                                                            file_filter)
+            return [str(chosen_file)]
+
+    def disable_loading(self):
+        self.disable_load_buttons()
+        self.loadingStarted.emit()
+
+    def enable_loading(self):
+        self.enable_load_buttons()
+        self.loadingFinished.emit()
```

Comment:
why do we need 2 different emits?

Reply:
The idea is that we may want to notify observers of this MVP that the loading has finished (more specific behaviour) or that the data has changed in some way (more general behaviour). 

Loading finished does not necessarily imply that the data has changed, and the data having changed does not necessarily imply that it was from a loading.

I hope this distinction will make dealing with the load widget a lot cleaner once it is embedded in a larger structure.

Reply:
why does loading finished not imply a change in data?

Reply:
It does, which is why both signals are emitted. But the converse is not necessarily true.

Reply:
could the slot for the data loaded signal just call the data changed methods (since we know they will always be called)? Then it is only one set of signals to follow.